File: pylox/interpreter.py
import logging

from pylox.expressions import BinaryExpr, UnaryExpr, LiteralExpr, GroupingExpr, Print, Expression, Var, Variable, Assign
from pylox.tokens import TokenType
from pylox.exceptions import RuntimeException
from pylox.environment import Environment

logger = logging.getLogger("pylox.interpreter")


class Interpreter(object):
    def __init__(self):
        self.environment = Environment()

    def interpret(self, statements):
        try:
            for stmt in statements:
                self._execute(stmt)
        except Exception as e:
            logger.error("Runtime Exception: {}".format(e))
            print("Runtime Exception: {}".format(e))

    def evaluate(self, expr):
        return expr.accept(self)

    def visit(self, expr):
        if isinstance(expr, BinaryExpr):
            return self._visit_binary_expr(expr)
        elif isinstance(expr, UnaryExpr):
            return self._visit_unary_expr(expr)
        elif isinstance(expr, LiteralExpr):
            return self._visit_literal_expr(expr)
        elif isinstance(expr, GroupingExpr):
            return self._visit_grouping_expr(expr)
        elif isinstance(expr, Print):
            return self._visit_print_stmt(expr)
        elif isinstance(expr, Expression):
            return self._visit_expression_stmt(expr)
        elif isinstance(expr, Var):
            return self._visit_var_stmt(expr)
        elif isinstance(expr, Variable):
            return self._visit_variable_expr(expr)
        elif isinstance(expr, Assign):
            return self._visit_assign_expr(expr)

    def _visit_literal_expr(self, expr):
        return expr.value

    def _visit_grouping_expr(self, expr):
        return self.evaluate(expr.expression)

    def _visit_unary_expr(self, expr):
        right = self.evaluate(expr.right)
        if expr.operator.token_type == TokenType.MINUS:
            self._check_number_operand(expr.operator, right)
            return -1 * right
        if expr.operator.token_type == TokenType.BANG:
            return not self._is_truthy(right)
        return None

    def _visit_binary_expr(self, expr):
        left = self.evaluate(expr.left)
        right = self.evaluate(expr.right)
        operator_type = expr.operator.token_type
        if operator_type == TokenType.GREATER:
            self._check_number_operands(expr.operator, left, right)
            return left > right
        if operator_type == TokenType.GREATER_EQUAL:
            self._check_number_operands(expr.operator, left, right)
            return left >= right
        if operator_type == TokenType.LESS:
            self._check_number_operands(expr.operator, left, right)
            return left < right
        if operator_type == TokenType.LESS_EQUAL:
            self._check_number_operands(expr.operator, left, right)
            return left <= right
        if operator_type == TokenType.MINUS:
            self._check_number_operands(expr.operator, left, right)
            return left - right
        if operator_type == TokenType.SLASH:
            self._check_number_operands(expr.operator, left, right)
            return left / right
        if operator_type == TokenType.STAR:
            self._check_number_operands(expr.operator, left, right)
            return left * right
        if operator_type == TokenType.PLUS:
            self._check_number_operands(expr.operator, left, right)
            return left + right
        if operator_type == TokenType.EQUAL_EQUAL:
            return self._is_equal(left, right)
        if operator_type == TokenType.BANG_EQUAL:
            return not self._is_equal(left, right)
        return None

    def _visit_expression_stmt(self, stmt):
        self.evaluate(stmt.expression)
        return None

    def _visit_print_stmt(self, stmt):
        value = self.evaluate(stmt.expression)
        print(value)
        return None

    def _visit_var_stmt(self, stmt):
        if stmt.initializer:
            value = self.evaluate(stmt.initializer)
        else:
            value = None
        self.environment.define(stmt.name.lexeme, value)
        return None

    def _visit_variable_expr(self, expr):
        return self.environment.get(expr.name)

    def _visit_assign_expr(self, expr):
        value = self.evaluate(expr.value)
        self.environment.assign(expr.name, value)
        return value

    def _execute(self, stmt):
        return stmt.accept(self)

    def _is_truthy(self, obj):
        if obj is None:
            return False
        if isinstance(obj, bool):
            return
        return True

    def _is_equal(self, obj1, obj2):
        if obj1 is None and obj2 is None:
            return True
        if obj1 is None:
            return False
        return obj1 == obj2

    def _check_number_operand(self, operator, operand):
        if type(operand) == float:
            return
        raise RuntimeException(operator, "Operand must be a number.")

    def _check_number_operands(self, operator, left, right):
        if type(left) == float and type(right) == float:
            return
        raise RuntimeException(operator, "Operands must be numbers.")

File: klox/src/main/kotlin/com/mab/lox/interpreter/interpreter.kt
package com.mab.lox.interpreter

import com.mab.lox.*
import com.mab.lox.TokenType.*
import com.mab.lox.stdlib.LoxStandardLib
import com.mab.lox.utils.isEqual
import com.mab.lox.utils.isTruthy
import com.mab.lox.utils.stringify

class Interpreter(
    private var environment: Environment = Environment()
) : Expr.Visitor<Any?>, Stmt.Visitor<Void?> {
    private val globals: Environment = Environment()
    private val locals: MutableMap<Expr, Int> = mutableMapOf()

    init {
        // Populate global functions from Lox's standard library.
        LoxStandardLib.global_functions.forEach { (name, function) ->
            globals.define(name, function)
        }
    }

    fun interpret(statements: List<Stmt>) {
        try {
            statements.forEach { execute(it) }
        } catch (error: RuntimeError) {
            Lox.runtimeError(error)
        }
    }

    private fun lookUpVariable(name: Token, expr: Expr): Any? {
        val distance = locals[expr]
        return if (distance != null) {
            environment.getAt(distance, name.lexeme)
        } else {
            globals.get(name)
        }
    }

    private fun evaluate(expr: Expr) = expr.accept(this)

    private fun execute(stmt: Stmt) = stmt.accept(this)

    fun resolve(expr: Expr, depth: Int) {
        locals[expr] = depth
    }

    private fun executeBlock(statements: List<Stmt>, environment: Environment) {
        val previous = this.environment
        try {
            this.environment = environment
            statements.forEach { execute(it) }
        } finally {
            this.environment = previous
        }
    }

    private fun checkNumberOperands(operator: Token, vararg operands: Any?) {
        operands.forEach {
            if (it !is Double && it !is Int) {
                throw RuntimeError(operator, "Operand(s) must be numbers.")
            }
        }
    }

    //
    // Expr.Visitor Interface Implementation.
    //

    override fun visitAssignExpr(expr: Expr.Assign): Any? {
        val value = evaluate(expr.value)
        val distance = locals[expr]
        if (distance != null) {
            environment.assignAt(distance, expr.name, value)
        } else {
            globals.assign(expr.name, value)
        }
        return value
    }

    override fun visitBinaryExpr(expr: Expr.Binary): Any? {
        val left = evaluate(expr.left)
        val right = evaluate(expr.right)

        when (expr.operator.type) {
            GREATER -> {
                checkNumberOperands(expr.operator, left, right)
                return when {
                    left is Double && right is Double -> left > right
                    left is Int && right is Int -> left > right
                    left is Double && right is Int -> left > right
                    left is Int && right is Double -> left > right
                    else -> throw RuntimeError(expr.operator, "Operands must be two numbers.")
                }
            }
            GREATER_EQUAL -> {
                checkNumberOperands(expr.operator, left, right)
                return when {
                    left is Double && right is Double -> left >= right
                    left is Int && right is Int -> left >= right
                    left is Double && right is Int -> left >= right
                    left is Int && right is Double -> left >= right
                    else -> throw RuntimeError(expr.operator, "Operands must be two numbers.")
                }
            }
            LESS -> {
                checkNumberOperands(expr.operator, left, right)
                return when {
                    left is Double && right is Double -> left < right
                    left is Int && right is Int -> left < right
                    left is Double && right is Int -> left < right
                    left is Int && right is Double -> left < right
                    else -> throw RuntimeError(expr.operator, "Operands must be two numbers.")
                }
            }
            LESS_EQUAL -> {
                checkNumberOperands(expr.operator, left, right)
                return when {
                    left is Double && right is Double -> left <= right
                    left is Int && right is Int -> left <= right
                    left is Double && right is Int -> left <= right
                    left is Int && right is Double -> left <= right
                    else -> throw RuntimeError(expr.operator, "Operands must be two numbers.")
                }
            }
            BANG_EQUAL -> return !isEqual(left, right)
            EQUAL_EQUAL -> return isEqual(left, right)
            MINUS -> {
                checkNumberOperands(expr.operator, right)
                return when {
                    left is Double && right is Double -> left - right
                    left is Int && right is Int -> left - right
                    left is Double && right is Int -> left - right
                    left is Int && right is Double -> left - right
                    else -> throw RuntimeError(expr.operator, "Operands must be two numbers.")
                }
            }
            PLUS -> {
                return when {
                    left is Double && right is Double -> left + right
                    left is Int && right is Int -> left + right
                    left is Double && right is Int -> left + right
                    left is Int && right is Double -> left + right
                    left is String && right is String -> left + right
                    else -> throw RuntimeError(expr.operator, "Operands must be two numbers or two strings.")
                }
            }
            SLASH -> {
                checkNumberOperands(expr.operator, left, right)
                return when {
                    left is Double && right is Double -> left / right
                    left is Int && right is Int -> left / right
                    left is Double && right is Int -> left / right
                    left is Int && right is Double -> left / right
                    else -> throw RuntimeError(expr.operator, "Operands must be two numbers.")
                }
            }
            STAR -> {
                checkNumberOperands(expr.operator, left, right)
                return when {
                    left is Double && right is Double -> left * right
                    left is Int && right is Int -> left * right
                    left is Double && right is Int -> left * right
                    left is Int && right is Int -> left * right
                    else -> throw RuntimeError(expr.operator, "Operands must be two numbers.")
                }
            }
            else -> {
                // Unreachable.
                return null
            }
        }
    }

    override fun visitCallExpr(expr: Expr.Call): Any? {
        val callee = evaluate(expr.callee)
        val arguments = mutableListOf<Any?>()
        expr.arguments.forEach {
            arguments.add(evaluate(it))
        }

        if (callee !is LoxCallable) {
            throw RuntimeError(expr.paren, "Can only call functions and classes.")
        }

        if (arguments.size != callee.arity()) {
            throw RuntimeError(expr.paren, "Expected ${callee.arity()} arguments but got ${arguments.size}.")
        }
        return callee.call(this, arguments)
    }

    override fun visitGetExpr(expr: Expr.Get): Any? {
        val obj = evaluate(expr.obj)
        if (obj is LoxInstance) return obj.get(expr.name)
        throw RuntimeError(expr.name, "Only instances have properties.")
    }

    override fun visitGroupingExpr(expr: Expr.Grouping): Any? = evaluate(expr.expression)

    override fun visitLiteralExpr(expr: Expr.Literal): Any? = expr.value

    override fun visitLogicalExpr(expr: Expr.Logical): Any? {
        val left = evaluate(expr.left)

        if (expr.operator.type == OR) {
            if (isTruthy(left)) return left
        } else {
            if (!isTruthy(left)) return left
        }

        return evaluate(expr.right)
    }

    override fun visitSetExpr(expr: Expr.Set): Any? {
        val obj = evaluate(expr.obj)
        if (obj !is LoxInstance) {
            throw RuntimeError(expr.name, "Only instances have fields.")
        }

        val value = evaluate(expr.value)
        obj.set(expr.name, value)
        return value
    }

    override fun visitSuperExpr(expr: Expr.Super): Any? {
        val distance = locals[expr]!!
        val superclass = environment.getAt(distance, "super") as LoxClass
        val obj = environment.getAt(distance - 1, "this") as LoxInstance
        val method = superclass.findMethod(expr.method.lexeme)
            ?: throw RuntimeError(expr.method, "Undefined property '${expr.method.lexeme}'.")
        return method.bind(obj)
    }

    override fun visitThisExpr(expr: Expr.This): Any? = lookUpVariable(expr.keyword, expr)

    override fun visitUnaryExpr(expr: Expr.Unary): Any? {
        val right = evaluate(expr.right)
        return when (expr.operator.type) {
            BANG -> isTruthy(right)
            MINUS -> {
                checkNumberOperands(expr.operator, right)
                when (right) {
                    is Double -> -right
                    is Int -> -right
                    else -> throw RuntimeError(expr.operator, "Operand must be a number.")
                }
            }
            else -> {
                // Unreachable.
                null
            }
        }
    }

    override fun visitVariableExpr(expr: Expr.Variable): Any? = lookUpVariable(expr.name, expr)

    //
    // Stmt.Visitor Interface Implementation.
    //

    override fun visitBlockStmt(stmt: Stmt.Block): Void? {
        executeBlock(stmt.statements, Environment(environment))
        return null
    }

    override fun visitClassStmt(stmt: Stmt.Class): Void? {
        var superclass: Any? = null
        if (stmt.superclass != null) {
            superclass = evaluate(stmt.superclass)
            if (superclass !is LoxClass) {
                throw RuntimeError(stmt.superclass.name, "Superclass must be a class.")
            }
        }

        environment.define(stmt.name.lexeme, null)
        if (stmt.superclass != null) {
            environment = Environment(environment)
            environment.define("super", superclass)
        }

        val methods = hashMapOf<String, LoxFunction>()
        stmt.methods.forEach { method ->
            val function = LoxFunction(method, environment, method.name.lexeme.equals("init"))
            methods[method.name.lexeme] = function
        }

        val klass = LoxClass(stmt.name.lexeme, superclass as LoxClass, methods)

        if (superclass != null) {
            // Pop the environment to remove the "super" definition.
            environment = environment.enclosing!!
        }

        environment.assign(stmt.name, klass)
        return null
    }

    override fun visitExpressionStmt(stmt: Stmt.Expression): Void? {
        evaluate(stmt.expression)
        return null
    }

    override fun visitFunctionStmt(stmt: Stmt.Function): Void? {
        val function = LoxFunction(stmt, environment, false)
        environment.define(stmt.name.lexeme, function)
        return null
    }

    override fun visitIfStmt(stmt: Stmt.If): Void? {
        if (isTruthy(evaluate(stmt.condition))) {
            execute(stmt.thenBranch)
        } else if (stmt.elseBranch != null) {
            execute(stmt.elseBranch)
        }
        return null
    }

    override fun visitPrintStmt(stmt: Stmt.Print): Void? {
        val value = evaluate(stmt.expression)
        println(stringify(value))
        return null
    }

    override fun visitReturnStmt(stmt: Stmt.Return): Void? {
        // If an expression follows the "return" keyword, evaluate the expr and throw it up the stack.
        val value = stmt.value?.let { evaluate(it) }
        throw Return(value)
    }

    override fun visitVarStmt(stmt: Stmt.Var): Void? {
        val value = stmt.initializer?.let { evaluate(it) }
        environment.define(stmt.name.lexeme, value)
        return null
    }

    override fun visitWhileStmt(stmt: Stmt.While): Void? {
        while (isTruthy(evaluate(stmt.condition))) {
            execute(stmt.body)
        }
        return null
    }
}
